### PR TITLE
add sub directory path to `--sarif-category` if it does not match `${WORKSPACE}`

### PR DESCRIPTION
**Pull request by arilivigni**
Check if `${PWD} == ${WORKSPACE}` if not then add relative `${PWD}` to `--sarif-category` option
This is implemented for both the Linux and Windows shared library `ExecuteCodeQL.groovy`

### CODE REVIEW

**Comment by arilivigni** [comment]
@lindluni updates in the PR to include Windows support

**Comment by lindluni** [comment]
Thank you @arilivigni, I will plug this into Jenkins and test it quick, but it looks good to me

**Comment by arilivigni** [comment]
@lindluni is the Jenkins testing down as part of the checks not comprehensive? Is there a test I can run through on the Jenkins myself to verify so I can make and test changes better in the future?

**Comment by lindluni** [comment]
It is comprehensive, but this scenario is not included since this requires configuration in the pipeline itself. Can you take a look at adding a second scenario to the vulnerable-node repos Jenkinsfile, it should be as simple as adding another  test but wrap it in the “dir” directive to change the working directory.

if you don’t get to it, I should have some time in a few hours.

**Comment by arilivigni** [comment]
@lindluni I can totally do that and test it. Thanks for the direction to validate.

**Comment by arilivigni** [comment]
@lindluni finally got this working and understand windows powershell paths more than I ever wanted to 🤣 
go in base vulnerable-go directory :: https://jenkins.github-apps.io/job/Go%20CodeQL%20-%20Windows/96/console
Javascript  vulnerable-go in public directory :: https://jenkins.github-apps.io/job/Go%20CodeQL%20-%20Windows/95/console

**Comment by arilivigni** [comment]
@lindluni when you get chance can you review or give feedback on other changes needed. 🙇

**Review by lindluni — CHANGES_REQUESTED** [review]
Hey @arilivigni in general, this looks good, thank you for the effort. I noted we need to copy the Windows absolute path changes to Linux as well (even ones I may have missed, they all need to be copied).

And then there is something weird going on with the Windows library we need to investigate, I provided the API call to pull the generated categories for review in the comments, but the code scanning part of it works like a charm, just this minor quirk, which has to be corrected, as we need to know when users are scanning the whole repo, vs only scanning subdirectories for compliance reasons.

Thank you again for taking this on, lets figure out this last quirk and get this over the hump.

**Comment by lindluni** [comment]
Also note I create two pipelines, one for Windows and one for Linux that runs both the regular and sub-directory scans in a single pipeline:

Windows: https://jenkins.github-apps.io/job/Go%20CodeQL%20-%20Windows/
Linux: https://jenkins.github-apps.io/job/Go%20CodeQL%20-%20Linux/

File: jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy [hunk 1]
```diff
@@ -121,14 +129,29 @@ def call(Org, Repo, Branch, Language, BuildCommand, Token, InstallCodeQL) {
         }
         Write-Output "Database initialized"
 
+        Write-Output "Check if the current directory matches \$Env:WORKSPACE"
```
[the remark this answers is not in the document]
Comment (arilivigni) [reply jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy:132]: I will look into it. The output contains the right category so it is the reporting that is messing with the sarif category?

File: jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy [hunk 2]
```diff
@@ -73,20 +76,25 @@ def call(Org, Repo, Branch, Language, BuildCommand, Token, InstallCodeQL) {
             Invoke-WebRequest -Method Get -OutFile "codeql.tgz" -Uri "https://github.com/github/codeql-action/releases/download/\$Id/codeql-bundle-win64.tar.gz"
 
             Write-Output "Extracting CodeQL archive"
-            tar -xf codeql.tgz --directory "\$Env:WORKSPACE"
+            tar -xf codeql.tgz --directory "\$pwd"
```
[the remark this answers is not in the document]
Comment (arilivigni) [reply jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy:79]: Done

File: jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy [hunk 3]
```diff
@@ -40,18 +40,21 @@ def call(Org, Repo, Branch, Language, BuildCommand, Token, InstallCodeQL) {
     }
 
     powershell """
-        \$json_file = ".github\\emass.json"
+        Write-Output "WORKSPACE: \$Env:WORKSPACE"
+        Write-Output "PWD: \$pwd"
 
-        Write-Output "Validating .github\\emass.json"
+        \$json_file = "\$Env:WORKSPACE\\.github\\emass.json"
```
[the remark this answers is not in the document]
Comment (arilivigni) [reply jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy:46]: Done

File: jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (arilivigni) [reply jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy:132]: I just did the query on my recent run and it is correct. Did you change anything or how would this fix itself? I will do some more runs as well.

```
{
    "ref": "refs/heads/main",
    "commit_sha": "c0eb76ba613f5bde9156b671c4d0df380368a716",
    "analysis_key": "(default)",
    "environment": "{}",
    "category": "ois-javascript-public",
    "error": "",
    "created_at": "2023-07-25T13:48:38Z",
    "results_count": 19,
    "rules_count": 201,
    "id": 103559041,
    "url": "https://api.github.com/repos/department-of-veterans-affairs/vulnerable-go/code-scanning/analyses/103559041",
    "sarif_id": "f12d7ab4-2af1-11ee-8130-d18f054e1cef",
    "tool": {
      "name": "CodeQL",
      "guid": null,
      "version": "2.14.0"
    },
    "deletable": true,
    "warning": ""
  },
  {
    "ref": "refs/heads/main",
    "commit_sha": "c0eb76ba613f5bde9156b671c4d0df380368a716",
    "analysis_key": "(default)",
    "environment": "{}",
    "category": "ois-go",
    "error": "",
    "created_at": "2023-07-25T13:44:27Z",
    "results_count": 6,
    "rules_count": 51,
    "id": 103557645,
    "url": "https://api.github.com/repos/department-of-veterans-affairs/vulnerable-go/code-scanning/analyses/103557645",
    "sarif_id": "5de363e0-2af1-11ee-912f-63c714c67ec6",
    "tool": {
      "name": "CodeQL",
      "guid": null,
      "version": "2.14.0"
    },
    "deletable": true,
    "warning": ""
  },
```

[the remark this answers is not in the document]
Comment (arilivigni) [reply jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy:132]: I think this is happening on the linux side not the Windows side. I am troubleshooting that now

[the remark this answers is not in the document]
Comment (arilivigni) [reply jenkins/shared-libraries/windows/vars/ExecuteCodeQL.groovy:132]:
```
WORKSPACE: /home/jenkins/workspace/workspace/Go CodeQL - Linux
PWD: /usr/local/jenkins-service
Check if PWD matches WORKSPACE
The current directory and /home/jenkins/workspace/workspace/Go CodeQL - Linux do NOT match.
Sarif Category: ois-go-jenkins-service
```

```
WORKSPACE: /home/jenkins/workspace/workspace/Go CodeQL - Linux
PWD: /usr/local/jenkins-service
Check if PWD matches WORKSPACE
The current directory and /home/jenkins/workspace/workspace/Go CodeQL - Linux do NOT match.
Sarif Category: ois-javascript-jenkins-service
```

**Comment by arilivigni** [comment]
@lindluni this is ready now with the linux changes applied along with the windows for the shared libraries 🚀

File: jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy [hunk 4]
```diff
@@ -47,25 +47,29 @@ def call(org, repo, branch, language, buildCommand, token, installCodeQL) {
     }
 
     sh '''
+        echo "WORKSPACE: ${WORKSPACE}"
```
Comment (lindluni) [inline jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy:50]: In a follow on PR can we update this to be natural language so users don’t send us requests trying to guess what these values mean:

```sh
echo “Current Jenkins Workspace: ${WORKSPACE}”
echo “Current working directory: ${PWD}”
echo “CodeQL database path: ${DATABASE_PATH}”
echo “CodeQL database bundle file name: ${DATABASE_BUNDLE}”
```

File: jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy [hunk 5]
```diff
@@ -147,11 +151,25 @@ def call(org, repo, branch, language, buildCommand, token, installCodeQL) {
         fi
         echo "Database initialized"
 
+        echo "WORKSPACE: ${WORKSPACE}"
+        echo "PWD: \${PWD}"
+        echo "Check if PWD matches WORKSPACE"
```
Comment (lindluni) [inline jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy:156]: Can we make these natural language as well in the follow on PR, the last line should be `Check if current working directory and Jenkins workspace are the same directory`

File: jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy [hunk 6]
```diff
@@ -147,11 +151,25 @@ def call(org, repo, branch, language, buildCommand, token, installCodeQL) {
...
+        if [ "\${PWD}" = "${WORKSPACE}" ]; then
+            echo "The current directory and ${WORKSPACE} match."
```
Comment (lindluni) [inline jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy:158]: And this one should just be `The current working directory and Jenkins workspace match` (also with no period)

File: jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy [hunk 7]
```diff
@@ -147,11 +151,25 @@ def call(org, repo, branch, language, buildCommand, token, installCodeQL) {
...
+            SUBDIR=''
+            SEP=''
+        else
+            echo "The current directory and ${WORKSPACE} do NOT match."
```
Comment (lindluni) [inline jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy:162]: This one should be `The current working directory and Jenkins workspace do not match, updating the SARIF category value to deduplicate Code Scanning results`

File: jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy [hunk 8]
```diff
@@ -147,11 +151,25 @@ def call(org, repo, branch, language, buildCommand, token, installCodeQL) {
...
+            SUBDIR=\$( echo \${PWD} | awk -F'/' '{print \$NF}' )
+            SEP='-'
+        fi
+        echo "Sarif Category: ois-${LANGUAGE}\${SEP}\${SUBDIR}"
```
Comment (lindluni) [inline jenkins/shared-libraries/linux/vars/ExecuteCodeQL.groovy:166]:
```sh
echo “The SARIF category has been configured to `<value here>`”
```

**Review by lindluni — APPROVED** [review]
Just a few comments on log statements to hopefully reduce confusion and limit support requests to understand what those log statements mean, but I am not going to hold up this PR. If you have cycles to a do a quick follow on PR to update the log statements that would be awesome, otherwise I will take care of it when I get back.

Note: I only commented on the log statements on the Linux library, the same statements should be copied to the Windows library.

Thank you @arilivigni for your work on this one and patience while we got it right!

**Pull request merged**